lib/block_view_lockups.go

```diff
@@ -818,109 +1174,541 @@ func (bav *UtxoView) _connectCoinLockup(
+					// remaining cases:
+					// First case:
+					// proposed:                       <--------------------->
+					// existing:
+					//
+					// Second case:
+					// proposed:                       <--------------------->
+					// existing:                                   <--------->
+					//
+					// The second case is fine to leave as it will be taken care of in the
+					// subsequent iteration. However, we make a special note to capture the
+					// remaining LockedBalanceEntry present in the first case.
+
+					// We check if there's another locked balance entry sometime in the future.
+					splitLockedBalanceEntry, remainingLockedBalanceEntry, err := SplitVestedLockedBalanceEntry(
+						proposedLockedBalanceEntry,
+						existingLockedBalanceEntry.UnlockTimestampNanoSecs,
+						existingLockedBalanceEntry.VestingEndTimestampNanoSecs)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(err, "_connectCoinLockup failed to compute vested split")
+					}
+
+					// Consolidate the split and existing locked balance entry.
+					combinedBalanceBaseUnits, err := SafeUint256().Add(
+						&splitLockedBalanceEntry.BalanceBaseUnits,
+						&existingLockedBalanceEntry.BalanceBaseUnits)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
+								"_connectCoinLockup")
+					}
+					splitLockedBalanceEntry.BalanceBaseUnits = *combinedBalanceBaseUnits
+
+					// Set the now combined splitLockedBalanceEntry into the view.
+					bav._setLockedBalanceEntry(splitLockedBalanceEntry)
+
+					// Update the proposed locked balance entry with the remaining portion.
+					proposedLockedBalanceEntry = remainingLockedBalanceEntry
+
+					// (3b-iii) On the final iteration, the remaining proposedLockedBalanceEntry
+					//          is the only vesting schedule left.
+					if ii == len(lockedBalanceEntries)-1 {
+						bav._setLockedBalanceEntry(proposedLockedBalanceEntry)
+					}
+				}
+
+				// (3b-iv) By now, we know the edges to be trimmed as best possible.
+				// We check if the existing and proposed now overlap perfectly in time and combine if so.
+				if (existingLockedBalanceEntry.UnlockTimestampNanoSecs ==
+					proposedLockedBalanceEntry.UnlockTimestampNanoSecs) &&
+					(existingLockedBalanceEntry.VestingEndTimestampNanoSecs ==
+						proposedLockedBalanceEntry.VestingEndTimestampNanoSecs) {
+					// Combine the remaining balance.
+					combinedBalanceBaseUnits, err := SafeUint256().Add(
+						&existingLockedBalanceEntry.BalanceBaseUnits,
+						&proposedLockedBalanceEntry.BalanceBaseUnits)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
+								"_connectCoinLockup")
+					}
+
+					// Update the remaining entry.
+					proposedLockedBalanceEntry.BalanceBaseUnits = *combinedBalanceBaseUnits
+					bav._setLockedBalanceEntry(proposedLockedBalanceEntry)
+				}
+			}
+		}
+	}
```

Comment:
I think this works, and it's really well-written. I really thought I'd be able to find a bug but... I couldn't so far. Great work with this. I'm still on the fence as to whether we want to do it this way vs the "capped unlock" way but will call you about it tomorrow to discuss. On the bright side, it looks like the "capped unlock" approach would be deleting a bunch of this code, except for the last part where you consolidate things with the same start+end time.